### PR TITLE
threaded dests: add `reopen()` with exponential backoff strategy

### PR DESCRIPTION
**Pull request by alltilla**

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build FAILURE

File: lib/exponential-backoff/exponential-backoff.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ *
+ */
+
+#ifndef EXPONENTIAL_BACKOFF_H
+#define EXPONENTIAL_BACKOFF_H
+
+#include "syslog-ng.h"
+
+#define EXPONENTIAL_BACKOFF_OPTIONS_INITIAL_SECONDS_DEFAULT (0.1)
+#define EXPONENTIAL_BACKOFF_OPTIONS_MAXIMUM_SECONDS_DEFAULT (1)
```
Comment (MrAnno) [inline lib/exponential-backoff/exponential-backoff.h:30]: This seems very low in my opinion compared to the old default. I think we shouldn't go lower than 10 seconds by default. If we are planning to introduce this for traditional syslog/network drivers too, I would also consider going with different defaults there (or just keep the old 60 seconds as maximum?).

Note: connection attempts are logged internally on a high log level. For example, using the HTTP produces 2 error messages with each reopen attempt:

```
curl: error sending HTTP request; url='...', error='Couldn\'t connect to server', worker_index='0', driver='#anon-destination0#1', location='...'
Server disconnected while preparing messages for sending, trying again; driver='#anon-destination0#1', location='...', worker_index='0', reopen='0.00', batch_size='1'
```

TLDR: Do we really want `1` as maximum? If so, then we should do something about internal logging connection attempts as we are currently spamming 2-3 error messages every second per worker.
In my opinion, we should rather choose a higher default (60 or 10 would be my suggestion) and keep the logging as-is. The exponential retry mechanism will make sure that we reconnect instantly in case of momentary errors, and we even have 5 more attempts before reaching `1` second of wait time and 10 attempts before reaching `60`.

File: lib/logthrdest/logthrdestdrv.c [hunk 2]
```diff
@@ -272,13 +280,31 @@ _rewind_batch(LogThreadedDestWorker *self)
   log_threaded_dest_worker_rewind_messages(self, self->batch_size);
 }
 
+static gdouble
+_get_next_reopen_seconds(LogThreadedDestWorker *self)
+{
+  if (self->time_reopen != -1)
```
Comment (MrAnno) [inline lib/logthrdest/logthrdestdrv.c:286]: This value is set explicitly by the following drivers:
- Kafka
- Python

File: lib/logthrdest/logthrdestdrv.c [hunk 3]
```diff
@@ -333,6 +359,7 @@ static void
 _process_result_success(LogThreadedDestWorker *self)
 {
   _accept_batch(self);
+  exponential_backoff_reset(self->exponential_backoff);
```
Comment (MrAnno) [inline lib/logthrdest/logthrdestdrv.c:362]: This is a double-reset here due to the function responsible for "explicit ack management".
It doesn't cause any problems, so I guess it's okay.

File: lib/exponential-backoff/exponential-backoff.h (hunk 1 above)
Comment (bazsi) [reply lib/exponential-backoff/exponential-backoff.h:30]: I assumed this was the "default" for max seconds, unless specified otherwise.
